File: makemodule.py
#!/bin/env python
"""
makemodule

Module generation tool

Copyright (c) 2015 Sam Saint-Pettersen.
Released under the MIT/X11 License.
"""
import sys
import os
import xml.dom.minidom as xml

class makemodule:

    def __init__(self, args):
        if len(args) == 1:
            self.displayUsage()
        else:
            self.writeModuleXML()

    def displayUsage(self):
        print(__doc__)
        print('Usage: makemodule [module..module]\n')
        sys.exit(1)

    def writeModuleXML(self):
        names = []
        enabled = []
        redirect = ''
        cleanup = False

        if os.name == 'nt':
            redirect = ' > a.tmp 2>&1'
            cleanup = True
        else:
            redirect = ' >> /dev/null 2>&1'

        for arg in sys.argv[1:]:
            names.append(arg)
            exitCode = int(os.system(arg + redirect))
            if exitCode == 32512:
                enabled.append(False)
            else:
                enabled.append(True)

        doc = xml.Document()
        c = doc.createElement('configuration')
        doc.appendChild(c)

        i = 0
        for name in names:
            m = doc.createElement('module')
            c.appendChild(m)
            n = doc.createElement('name')
            m.appendChild(n)
            n_is = doc.createTextNode(name)
            n.appendChild(n_is)
            e = doc.createElement('enabled')
            m.appendChild(e)
            e_is = doc.createTextNode(str(enabled[i]))
            e.appendChild(e_is)
            i = i + 1

        print('Writing modules.xml...')
        f = open('modules.xml', 'w')
        f.write(doc.toprettyxml())
        f.close()

        if os.name == 'nt': os.remove('a.tmp')

makemodule(sys.argv)

File: makemodule/MakeModule.cs
using System;
using System.ComponentModel;
using System.Collections;
using System.Text;
using System.Diagnostics;
using System.IO;
using System.Xml;

namespace makemodule
{
    class MakeModule
    {
        private static ArrayList names;
        private static ArrayList enabled;

        static void Main(string[] args)
        {
            names = new ArrayList();
            enabled = new ArrayList();

            if (args.Length == 0)
            {
                Console.WriteLine("\nmakemodule utility");
                Console.WriteLine("Copyright (c) 2015 Sam Saint-Pettersen.");
                Console.WriteLine("Released under the MIT/X11 License.");
                Console.WriteLine("\nmakemodule [module..module]");
                Environment.Exit(1);
            }

            foreach(string arg in args)
            {
                names.Add(arg);

                Process p = new Process();
                p.StartInfo.FileName = arg;
                p.StartInfo.UseShellExecute = false;
                p.StartInfo.RedirectStandardError = true;
                p.StartInfo.RedirectStandardOutput = true;
                p.StartInfo.CreateNoWindow = true;

                try
                {
                    p.Start();

                    p.WaitForExit();
                    if (p.ExitCode != 0) enabled.Add(true);
                    else enabled.Add(false);
                }
                catch(Win32Exception)
                {
                    enabled.Add(false);
                }
            }

            if(names.Count > 0) writeModuleXML();
        }

        static void writeModuleXML()
        {
            XmlDocument modulesDoc = new XmlDocument();
            XmlNode root = modulesDoc.CreateElement("configuration");
            modulesDoc.AppendChild(root);

            for(int i = 0; i < names.Count; i++)
            {
                XmlNode moduleNode = modulesDoc.CreateElement("module");
                XmlNode nameNode = modulesDoc.CreateElement("name");
                nameNode.InnerText = names[i].ToString();
                XmlNode enabledNode = modulesDoc.CreateElement("enabled");
                enabledNode.InnerText = enabled[i].ToString();
                moduleNode.AppendChild(nameNode);
                moduleNode.AppendChild(enabledNode);
                root.AppendChild(moduleNode);
            }

            Console.WriteLine("Writing modules.xml...");
            modulesDoc.Save("modules.xml");

            StreamReader sr = new StreamReader("modules.xml");
            ArrayList lines = new ArrayList();
            while(!sr.EndOfStream)
            {
                lines.Add(sr.ReadLine());
            }
            sr.Close();
            StreamWriter sw = new StreamWriter("modules.xml", false, Encoding.UTF8);
            sw.WriteLine("<?xml version=\"1.0\" encoding=\"utf-8\" ?>");
            foreach(string line in lines)
            {
                sw.WriteLine(line);
            }
            sw.Close();
        }
    }
}